### PR TITLE
Mostly fix Julia 1.0 compatibility

### PR DESCRIPTION
**Pull request by iht**
I have done some changes to bump this package to Julia 1.0.

I have translated all the `type` declarations into `mutable struct`, so instances of those types can be modified. Maybe some of those can be declared just `struct`. I am not very familiar with the code, so I am not sure of how each one of those types are used.

Lots of packages depend on this one, and are currently uninstallable on Julia 1.0; merging this PR will make it possible to install those packages.

### CODE REVIEW

**Comment by iht** [comment]
Ok. I should have checked the list of open PRs before jumping into fixing something. Please feel free to drop this PR if #41 is accepted.

File: .travis.yml [hunk 1]
```diff
@@ -3,17 +3,16 @@ os:
     - linux
     - osx
 julia:
-    - 0.4
-    - 0.5
-    - 0.6
+    - 0.7
+    - 1.0
```
Comment (omus) [inline .travis.yml]: REQUIRE file and appveyor.yml also need to be updated

Comment (iht) [reply .travis.yml]: Done

File: appveyor.yml [hunk 2]
```diff
@@ -1,11 +1,9 @@
...
-  - JULIA_URL: "https://julialang-s3.julialang.org/bin/winnt/x86/0.4/julia-0.4-latest-win32.exe"
-  - JULIA_URL: "https://julialang-s3.julialang.org/bin/winnt/x64/0.4/julia-0.4-latest-win64.exe"
-  - JULIA_URL: "https://julialang-s3.julialang.org/bin/winnt/x86/0.5/julia-0.5-latest-win32.exe"
-  - JULIA_URL: "https://julialang-s3.julialang.org/bin/winnt/x64/0.5/julia-0.5-latest-win64.exe"
-  - JULIA_URL: "https://julialang-s3.julialang.org/bin/winnt/x86/0.6/julia-0.6-latest-win32.exe"
-  - JULIA_URL: "https://julialang-s3.julialang.org/bin/winnt/x64/0.6/julia-0.6-latest-win64.exe"
+  - JULIA_URL: "https://julialang-s3.julialang.org/bin/winnt/x86/0.7/julia-0.7-latest-win32.exe"
+  - JULIA_URL: "https://julialang-s3.julialang.org/bin/winnt/x64/0.7/julia-0.7-latest-win64.exe"
+  - JULIA_URL: "https://julialang-s3.julialang.org/bin/winnt/x86/1.0/julia-1.0-latest-win32.exe"
+  - JULIA_URL: "https://julialang-s3.julialang.org/bin/winnt/x64/1.0/julia-1.0-latest-win64.exe"
```
Comment (omus) [inline appveyor.yml]: You'll probably run into problems with this. You can use the template in https://github.com/JuliaCI/Appveyor.jl

Comment (iht) [reply appveyor.yml]: Thanks for the hint, new commit pushed

File: appveyor.yml [hunk 3]
```diff
@@ -21,19 +20,13 @@ notifications:
...
-# Run installer silently, output to C:\projects\julia
-  - C:\projects\julia-binary.exe /S /D=C:\projects\julia
+  - ps: iex ((new-object net.webclient).DownloadString("https://raw.githubusercontent.com/JuliaCI/Appveyor.jl/version-1/bin/install.ps1"))
 
 build_script:
 # Need to convert from shallow to complete for Pkg.clone to work
   - IF EXIST .git\shallow (git fetch --unshallow)
-  - C:\projects\julia\bin\julia -e "versioninfo();
+  - C:\julia\bin\julia -e "versioninfo(); import Pkg;
       Pkg.clone(pwd(), \"LibCURL\"); Pkg.build(\"LibCURL\")"
```
Comment (omus) [inline appveyor.yml]: `versioninfo()` isn't going to work anymore. `InteractiveUtils.versioninfo()` will work.

File: src/LibCURL.jl [hunk 4]
```diff
@@ -11,25 +11,25 @@ const curl_off_t = Int64
...
+curl_easy_setopt(handle, opt, ptrval::Ptr{T}) where {T} = ccall((:curl_easy_setopt, libcurl), CURLcode, (Ptr{CURL}, CURLoption, Ptr{T}), handle, opt, ptrval)
 curl_easy_setopt(handle, opt, ptrval::AbstractString) = ccall((:curl_easy_setopt, libcurl), CURLcode, (Ptr{CURL}, CURLoption, Ptr{UInt8}), handle, opt, ptrval)
 
-curl_multi_setopt(handle, opt, ptrval::Array) = ccall((:curl_multi_setopt, libcurl), CURLMcode, (Ptr{CURLM}, CURLMoption, Ptr{Void}), handle, opt, pointer(ptrval))
+curl_multi_setopt(handle, opt, ptrval::Array) = ccall((:curl_multi_setopt, libcurl), CURLMcode, (Ptr{CURLM}, CURLMoption, Ptr{Cvoid}), handle, opt, pointer(ptrval))
 curl_multi_setopt(handle, opt, ptrval::Integer) = ccall((:curl_multi_setopt, libcurl), CURLMcode, (Ptr{CURLM}, CURLMoption, Clong), handle, opt, ptrval)
-curl_multi_setopt{T}(handle, opt, ptrval::Ptr{T}) = ccall((:curl_multi_setopt, libcurl), CURLMcode, (Ptr{CURLM}, CURLMoption, Ptr{T}), handle, opt, ptrval)
+curl_multi_setopt(handle, opt, ptrval::T) where {T} = ccall((:curl_multi_setopt, libcurl), CURLMcode, (Ptr{CURLM}, CURLMoption, T), handle, opt, ptrval)
```
Comment (omus) [inline src/LibCURL.jl]: This should be `Ptr{T}`

File: src/lC_common_h.jl [hunk 5]
```diff
@@ -51,32 +51,32 @@ type curl_fileinfo
...
+@ctypedef curl_read_callback Ptr{Cvoid}
 # enum curlsocktype
 const CURLSOCKTYPE_IPCXN = 0
 const CURLSOCKTYPE_LAST = 1
 # end
 @ctypedef curlsocktype Int32
-@ctypedef curl_sockopt_callback Ptr{Void}
-type curl_sockaddr
+@ctypedef curl_sockopt_callback Ptr{Cvoid}
+struct curl_sockaddr
```
Comment (omus) [inline src/lC_common_h.jl]: Should be a `mutable struct`

File: src/lC_curl_h.jl [hunk 6]
```diff
@@ -3,21 +3,21 @@
 
 @c Cint curl_strequal (Ptr{UInt8}, Ptr{UInt8}) libcurl
 @c Cint curl_strnequal (Ptr{UInt8}, Ptr{UInt8}, size_t) libcurl
-@c CURLFORMcode curl_formadd (Ptr{Ptr{Void}}, Ptr{Ptr{Void}}) libcurl
-@c Cint curl_formget (Ptr{Void}, Ptr{Void}, curl_formget_callback) libcurl
-@c Void curl_formfree (Ptr{Void},) libcurl
+@c CURLFORMcode curl_formadd (Ptr{Ptr{Cvoid}}, Ptr{Ptr{Nothing}}) libcurl
+@c Cint curl_formget (Ptr{Cvoid}, Ptr{Nothing}, curl_formget_callback) libcurl
+@c Cvoid curl_formfree (Ptr{Nothing},) libcurl
```
Comment (omus) [inline src/lC_curl_h.jl]: These `Nothing` types should be `Cvoid`

**Comment by omus** [comment]
Thanks @iht for getting the ball rolling. I've added some additional changes and I noticed that we still need to create a BinaryBuilder repo for things to work correctly on 1.0. I'll be taking care of that now.

**Comment by iht** [comment]
Thanks @omus, happy to help with anything I could contribute. I think all the changes you have requested are now in this pull request. Looking forward to having 1.0 compatibility :).

**Comment by scls19fr** [comment]
I wonder if this Travis job shouldn't re-triggered https://travis-ci.org/JuliaWeb/LibCURL.jl/jobs/414573418

**Comment by omus** [comment]
Waiting on: https://github.com/JuliaWeb/LibCURLBuilder/pull/2

**Comment by omus** [comment]
Progress! Should have this PR updated and working shortly

**Comment by omus** [comment]
The BinaryBuilder update has run into a roadblock at the moment. I'll see if I can wrap this up with Homebrew and WinRPM tomorrow to get a working version of LibCURL.jl out. Sorry for the delay.

**Comment by omus** [comment]
It looks like WinRPM has Julia 1.0 issues. Once those are addressed this PR should be working. I'll get this merged so that at least the master branch of this package will be ready to go.

**Pull request merged**